src/drivers/esp8266_drv.c

```diff
@@ -31,82 +31,150 @@
-#define _SERIAL_CMD_MAX_LEN	1024
-#define _SERIAL_PARITY		0
-#define _SERIAL_RX_BUFF_SIZE	256
-#define _SERIAL_STOP_BITS	1
-#define _SERIAL_TX_BUFF_SIZE	256
-#define _TASK_STACK_SIZE	512
-#define _TASK_THREAD_NAME	"ESP8266 Driver"
-
-enum _cmd {
-        _CMD_UNKNOWN = 0,
-        _CMD_NOOP,
-        _CMD_INIT,
-        _CMD_CLIENT_AP_GET,
-        _CMD_CLIENT_AP_SET,
-        _CMD_CLIENT_IP_GET,
-        _CMD_DAEMON_SETUP,
+#define AT_TASK_TIMEOUT_MS	3
+#define CHECK_DONE_SLEEP_MS	3600000
+#define CLIENT_BACKOFF_MS	30000
+#define INIT_FAIL_SLEEP_MS	10000
+#define LOG_PFX			"[ESP8266 Driver] "
+#define MAX_CHANNELS		5
+#define SERIAL_BAUD		115200
+#define SERIAL_BITS		8
+#define SERIAL_CMD_MAX_LEN	1024
+#define SERIAL_PARITY		0
+#define SERIAL_RX_BUFF_SIZE	256
+#define SERIAL_STOP_BITS	1
+#define SERIAL_TX_BUFF_SIZE	256
+#define TASK_STACK_SIZE		512
+#define TASK_THREAD_NAME	"ESP8266 Driver"
+
+struct device {
+        struct Serial *serial;
+        enum dev_init_state init_state;
 };
 
-struct _client {
-        struct esp8266_client_info info;
+struct client {
         const struct wifi_client_cfg *config;
-        tiny_millis_t next_check_ms;
+        struct esp8266_client_info info;
+        tiny_millis_t next_join_attempt;
 };
 
-struct _daemon {
+struct daemon {
         bool listening;
 };
 
-struct _channel {
+struct channel {
         struct Serial *serial;
         size_t tx_chars_buffered;
         bool in_use;
         bool connected;
+        bool created_externally;
 };
 
-static struct _state {
-        struct Serial *serial;
+struct comm {
         new_conn_func_t *new_conn_cb;
-        struct _channel channels[_MAX_CHANNELS];
-        struct _client client;
-        struct _daemon daemon;
-        bool cmd_ip;
-        enum _cmd cmd;
-        enum dev_init_state dev_state;
-        tiny_millis_t cmd_sleep_until;
+        struct channel channels[MAX_CHANNELS];
+};
+
+/**
+ * These enums define the various checks that we can perform.
+ * Checks can lead to action.  Actions lead to updated state,
+ * which can then trigger checks.  We go round and round with
+ * this methodology to ensure that all decision making is kept
+ * in a single location.  Checks stop when nothing needs to
+ * be changed.
+ *
+ * Note that each check represents a tiny state machine.  These
+ * state machines have interdependeinces upon one another. In
+ * example the client wifi (CHECK_CLIENT) requires that the wifi
+ * subsystem (CHECK_INIT) be initialized. Thus we need a priority
+ * with the lowest value being the highest priority.
+ */
+enum check {
+        CHECK_INIT,
+        CHECK_CLIENT,
+        CHECK_DAEMON,
+        CHECK_DATA,
+        __NUM_CHECKS, /* Always the last */
+};
+
+struct cmd {
+        bool in_progress;
+        tiny_millis_t check[__NUM_CHECKS];
+};
+
+static struct {
+        struct device device;
+        struct client client;
+        struct daemon daemon;
+        struct comm comm;
+        struct cmd cmd;
 } state;
```

Comment:
Chap 4 of the Kernel style guide suggests that C is a spartan language, but global variables should be descriptive. While state isn't a global variable, it's scope is in that global direction which I think supports the more explicit naming of the variable and help new people understand it better.